File: Problems/gloomb_sequence.py
MAX = 100001
arr = [0]*MAX
def GolombSequence(N): 
  
  
    cnt = 0
  
    arr[0] = 0
    arr[1] = 1
  
    M = dict() 
  
    M[2] = 2
  
    for i in range(2, N + 1): 
  
        if (cnt == 0): 
            arr[i] = 1 + arr[i - 1] 
            cnt = M[arr[i]] 
            cnt -= 1

        else: 
            arr[i] = arr[i - 1] 
            cnt -= 1
  
        M[i] = arr[i] 
   
  
T = int(input())

for i in range(T):
    l = list(map(int,input().split())) 
    L = l[0] 
    R = l[1] 
    GolombSequence(R)
    s = 0
    for i in range(L,R+1):
        s+=(arr[i]*arr[i])
    print(s)

File: Problems/gloomb_sequence.cpp
#include<bits/stdc++.h>
#define MAX 100000000000
using namespace std; 
long long GolombSequence(long long L,long long R) 
{ 
  
    long long arr[MAX]; 

    long long cnt = 0; 

    arr[0] = 0; 
    arr[1] = 1; 
    map<long long,long long> M; 
  
    M[2] = 2; 

    for (long long i = 2; i <= R; i++) { 
        if (cnt == 0) { 
            arr[i] = 1 + arr[i - 1]; 
            cnt = M[arr[i]]; 
            cnt--; 
        } 
        else { 
            arr[i] = arr[i - 1]; 
            cnt--; 
        } 
  
        M[i] = arr[i]; 
    } 
    long long sum=0;
    for(long long i=L;i<=R;i++)
    {
        cout<<arr[i]<<"  ";
    }

    return sum%1000000007;
} 
  
int main() 
{ 
    long long T;
    long long L,R;
    cin>>T;
    while(T--)
    {
        cin>>L>>R;
        cout<<GolombSequence(L,R)<<endl;
    }
    return 0; 
} 